src/mcts/search.cc

```diff
@@ -609,191 +338,505 @@ void Search::UpdateRemainingMoves() {
   if (remaining_playouts_ <= 1) remaining_playouts_ = 1;
 }
 
-void Search::ExtendNode(Node* node, const PositionHistory& history) {
-  // Not taking mutex because other threads will see that N=0 and N-in-flight=1
-  // and will not touch this node.
-  const auto& board = history.Last().GetBoard();
-  auto legal_moves = board.GenerateLegalMoves();
+std::pair<Move, Move> Search::GetBestMove() const {
+  SharedMutex::SharedLock lock(nodes_mutex_);
+  Mutex::Lock counters_lock(counters_mutex_);
+  return GetBestMoveInternal();
+}
 
-  // Check whether it's a draw/lose by rules.
-  if (legal_moves.empty()) {
-    // Checkmate or stalemate.
-    if (board.IsUnderCheck()) {
-      // Checkmate.
-      node->MakeTerminal(GameResult::WHITE_WON);
+std::pair<Move, Move> Search::GetBestMoveInternal() const
+    REQUIRES_SHARED(nodes_mutex_) REQUIRES_SHARED(counters_mutex_) {
+  if (responded_bestmove_) return best_move_;
+  if (!root_node_->HasChildren()) return {};
+
+  float temperature = kTemperature;
+  if (temperature && kTempDecayMoves) {
+    int moves = played_history_.Last().GetGamePly() / 2;
+    if (moves >= kTempDecayMoves) {
+      temperature = 0.0;
     } else {
-      // Stalemate.
-      node->MakeTerminal(GameResult::DRAW);
+      temperature *=
+          static_cast<float>(kTempDecayMoves - moves) / kTempDecayMoves;
     }
-    return;
   }
 
-  // If it's root node and we're asked to think, pretend there's no draw.
-  if (node != root_node_) {
-    if (!board.HasMatingMaterial()) {
-      node->MakeTerminal(GameResult::DRAW);
-      return;
-    }
+  Node* best_node = temperature && root_node_->GetN() > 1
+                        ? GetBestChildWithTemperature(root_node_, temperature)
+                        : GetBestChild(root_node_);
 
-    if (history.Last().GetNoCapturePly() >= 100) {
-      node->MakeTerminal(GameResult::DRAW);
-      return;
+  Move ponder_move;
```

Comment:
ponder_move never initialized before return?

Reply:
All members of it have initializers, so it will be initialized to a1a1 (aka nullmove)

Reply:
I hope it's ignored downstream then!

Reply:
I think just a comment will do fine, like `Move ponder_move; // uninitialized ponder_moves are ignored`